Add spec changes for image extensions (#298)
buildpacks/spec: buildpack.md

The lifecycle program uses buildpacks to build software artifacts from source code and pack the result into an OCI image.

This is accomplished in six phases:

1. **Analysis**, where metadata about OCI layers generated during a previous build are made available to buildpacks.
2. **Detection**, where an optimal selection of compatible buildpacks is chosen.

> 3. **Build-Ext** (optional, see the [Image Extension Interface](image-extension.md)), where image extensions generate Dockerfiles that can be used to extend the build and/or run base images.

Author: `Build-Ext` is not the best name... I like `Generate` because the purpose of this phase is to **generate** Dockerfiles.

Reviewer: I like "Generation"

Reviewer 2: If we call it Generation, then having the extension script become /bin/generate (rather than /bin/build) would be nice & consistent.

Author: Updated in bf837f6

| `<app>/.profile`       | [†](README.md#linux-only) Bash-formatted script sourced by shell before launch
| `<app>/.profile.bat`   | [‡](README.md#windows-only) BAT-formatted script sourced by shell before launch

> TODO: move this after analyze

Author: Doing this now would make the diff impossible to decipher.

Author: Actually... this isn't needed. What we are calling "analysis" all happens in the restore phase now.

## Phase #1: Detection

![Detection](img/detection.svg)

### Purpose

> The purpose of detection is to find an ordered group of image extensions and buildpacks to use during the build phase.

Author: ```suggestion
The purpose of detection is to find an ordered group of image extensions and buildpacks to use during the build-ext and build phases.
```

### Purpose

The purpose of detection is to find an ordered group of image extensions and buildpacks to use during the build-ext and build phases.

> These image extensions and buildpacks must be compatible with the app.

Reviewer: ```suggestion
```

I think we should remove this phrase, given that the `must` suggests a requirement, and given that "compatible" isn't well-defined.

For detect requirements that are specific to image extensions, see the [Image Extension Interface](image-extension.md).

### Process

**GIVEN:**
- An ordered list of image extension or buildpack groups resolved into buildpack implementations as described in [Order Resolution](#order-resolution)
- A directory containing application source code
- A shell, if needed,

> For each buildpack implementation in each group in order, the lifecycle MUST execute `/bin/detect`. Image extensions and buildpacks are both buildpack implementations. For simplicity, the following section will refer to both as "buildpacks". Image extensions MUST always be optional during detection.

Reviewer: Can we stop using the term "buildpack implementation"? Or at least swap the word order to "implementation buildpack" if we can't think of a better term for a non-meta-buildpack? Meta-buildpacks are still implementations of buildpacks, so I think it's difficult to understand these sections.

This could be a separate PR, but I feel like adding image extensions makes this even more confusing.

Some name ideas:
- atomic buildpacks
- executable buildpacks
- concrete buildpacks

Author: Please see https://github.com/buildpacks/spec/pull/300 - I went with "executable buildpack" because that seemed most natural according to the definition I used.

         **Then** the lifecycle MUST proceed to the next group or fail detection completely if no more groups are present.

      2. **If** at least one exit status from `/bin/detect` in the group is zero \

>          **Then** the lifecycle MUST select this group and proceed to the builder-ext phase.

Reviewer 3: ```suggestion
         **Then** the lifecycle MUST select this group and MAY proceed to the build-ext phase.
```

The `build-ext` phase is option right? So maybe this should be a `MAY`?

#### Order Resolution

During detection, an order definition for image extensions (if present) and an order definition for buildpacks MUST be resolved into individual buildpack implementations.

> Order definitions for image extensions MUST NOT contain nested orders. If an order definition for image extensions is present, it will be pre-pended to the order definition for buildpacks (as if the order definition for image extensions were an order buildpack).

Author: This sentence (while "accurate", I think) doesn't make much sense... I am not sure how to improve it without re-working the entire section.

Reviewer 2: "During detection, order definitions MUST be resolved into individual buildpack implementations. If present, Image Extensions MAY provide an order definition. Buildpacks MUST provide an order definition."

Not sure if that covers the intent.. which itself kinda hints the original still has wiggle room.. (It's possible in my rephrasing to interpret that each extension would supply an order definition, or that there would be one for all extensions). 

The original leaves bare the possibilty that an order defintiion for extensions is required, although the extensions themselves remain optional (as the binding of the 'if present' can be made to 'image extensions' rather than 'order definition').

#### Order Resolution

> During detection, an order definition for image extensions (if present) and an order definition for buildpacks MUST be resolved into individual buildpack implementations.

Reviewer: ```suggestion
During detection, order definitions for image extensions and buildpacks MUST be resolved into a single order definition containing only atomic buildpacks.
```

> Order definitions for image extensions MUST NOT contain nested orders. If an order definition for image extensions is present, it will be pre-pended to the order definition for buildpacks (as if the order definition for image extensions were an order buildpack).

Reviewer: ```suggestion
Order definitions for image extensions MUST NOT contain nested orders.
If an order definition for image extensions is present, it MUST be prepended to the order definition for buildpacks before the resolution process occurs.
```

To make this more clear, I think we should define the following six terms early in the spec:
1. Order buildpack
2. Atomic buildpack
3. Order image extension (not currently allowed)
4. Atomic image extension
5. Order build module (currently only order buildpack)
6. Atomic build module (atomic buildpacks and image extensions)

The names are just suggestions.
Maybe it's easier to say that image extensions are buildpacks?
1. Normal order buildpack
2. Normal atomic buildpack
3. Image extension order bulidpack (not currently allowed)
4. Image extension atomic buildpack
5. Order buildpack (only normal buildpacks allowed)
6. Atomic buildpack (atomic normal and extension buildpacks)

Author: @sclevine what is a build module? What is the difference between an order build module and an atomic build module?

Author: Let's hash this out in https://github.com/buildpacks/spec/pull/300/files - we can rebase this PR when that one is merged.

After analysis, the lifecycle MUST proceed to the build phase.

> ## Phase #3: Build-Ext (optional)

Author: This could potentially be moved to image-extension.md

Reviewer 3: I'm +1 on that

Author: Updated in b677794

### Purpose

The purpose of the build-ext phase is to generate Dockerfiles that can be used to extend the build and/or run base images. The Build-Ext phase MUST NOT be run for Windows builds.

### Process

**GIVEN:**
- The final ordered group of image extensions determined during the detection phase,
- A directory containing application source code,
- The Buildpack Plan,

> - An `<output>` directory used to store generated artifacts

Author: ```suggestion
- An `<output>` directory used to store generated artifacts,
```

Correspondingly, each `/bin/build` executable:

- MAY read from the `<app>` directory.
- MUST NOT write to the `<app>` directory.
- MAY read the build environment as described in the [Environment](#environment) section.
- MAY read the Buildpack Plan.
- MAY log output from the build process to `stdout`.
- MAY emit error, warning, or debug messages to `stderr`.
- MAY write any combination of Dockerfile, build.Dockerfile, and run.Dockerfile to the `<output>` directory. These files MUST adhere to the requirements listed below.
- MAY write key-value pairs to `<output>/launch.toml` that are provided as build args to Dockerfile or run.Dockerfile when extending the run image.
- MAY write key-value pairs to `<output>/build.toml` that are provided as build args to Dockerfile or build.Dockerfile when extending the build image.
- SHOULD write SBOM (Software-Bill-of-Materials) files as described in the [Software-Bill-of-Materials](#software-bill-of-materials) section describing any contributions to the app image or build environment.

#### Dockerfile Requirements

run.Dockerfiles have the following requirements:

> - TODO: list run.Dockerfile requirements

Author: We should add the instructions that are allowed

- Dockerfiles that target the run image (only) may indicate any image in their `FROM` instruction.

Dockerfiles and build.Dockerfiles have the following additional requirements:

- TODO: list build.Dockerfile requirements

> ## Phase #4: Extension

Author: This could potentially be moved to image-extension.md

### Purpose

The purpose of the extension phase is to apply the Dockerfiles generated in the build-ext phase to the build and/or run base images. The Extension phase MUST NOT be run for Windows builds.

### Process

The extension phase MUST be invoked separately for each base image (build or run) that is extended.

**GIVEN:**
- The final ordered group of Dockerfiles generated during the build-ext phase,

> - Buildpack-provided build args specified during the build-ext phase,

Author: ```suggestion
- Image extension-provided build args specified during the build-ext phase,
```

- A UUID provided as a `build_id` build arg,
- A user id provided as a `user_id` build arg, and a group id provided as a `group_id` build arg,

> - A Dockerfile build context provided as a `base_image` build arg,

Author: ```suggestion
- A Dockerfile FROM image provided as a `base_image` build arg,
```

The `build_id` build arg allows the Dockerfile to invalidate the cache after a certain layer. When the `$build_id` build arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).

> The `user_id` and `group_id` build args allow the Dockerfile to reset the `USER` to its original value if the user needs to be changed (e.g., to `root`) in order to perform privileged operations.

Author: ```suggestion
The `user_id` and `group_id` build args allow the Dockerfile to reset the image config's `User` field to its original value if the user needs to be changed (e.g., to `root`) during the build in order to perform privileged operations.
```
